Adding NAppsDirListener to observe napps changes

macartur opened the pull request:
Fix  #476

macartur committed:
Adding NAppsDirListener to observe napps changes

Updated CHANGELOG.rst
Fix  #476

cemsbr commented on kytos/core/controller.py:
```diff
@@ -205,6 +209,7 @@ def start_controller(self):
 
         self.log.info("Loading Kytos NApps...")
         self.load_napps()
+        self.napp_dir_listener.start()
```
To avoid concurrency problems, start it before load_napps().

cemsbr commented on kytos/core/controller.py:
```diff
@@ -595,6 +601,8 @@ def unload_napp(self, username, napp_name):
             napp_name (str): Name of the NApp to be unloaded.
         """
         napp = self.napps.pop((username, napp_name), None)
+        self.log.info(napp)
```
We already have log messages in the `if else` below. Is this one necessary?

cemsbr commented on kytos/core/napps/napps_dir_listener.py:
```diff
@@ -0,0 +1,67 @@
+            tuple: Tuple with username and napp_name.
+        """
+        relative_path = absolute_path.replace(self.napps_path, '')
+        log.info(relative_path)
```
I believe this is a debugging message.

cemsbr commented on kytos/core/napps/napps_dir_listener.py:
```diff
@@ -0,0 +1,67 @@
+                observer.
+        """
+        napp = self._get_napp(event.src_path)
+        self._controller.load_napp(*napp)
```
Log as info that the napp was enabled before calling controller.

cemsbr commented on kytos/core/napps/napps_dir_listener.py:
```diff
@@ -0,0 +1,67 @@
+                observer.
+        """
+        napp = self._get_napp(event.src_path)
+        self._controller.unload_napp(*napp)
```
Log as info that the napp was disabled before calling controller.

diraol commented on CHANGELOG.rst:
```diff
@@ -7,6 +7,8 @@ All notable changes to the kytos project will be documented in this file.
 ********************************
 Added
 =====
+- Added NAppDirListener class to manager napps when enabled or disabled by
```
`NAppDirListener to listen to NApps that were enabled/disabled by kytos-utils and then load/unload them on the current running controller.`

diraol commented on kytos/core/controller.py:
```diff
@@ -102,6 +103,9 @@ def __init__(self, options=None):
 
         self.register_kytos_endpoints()
 
+        #: Observer that handle napps changes.
```
`Observer that handle NApps when they are enabled or disabled.`

diraol commented on kytos/core/napps/napps_dir_listener.py:
```diff
@@ -0,0 +1,68 @@
+                observer.
+        """
+        napp = self._get_napp(event.src_path)
+        log.info('The NApp "%s/%s" was enabled.', *napp)
```
I don't think we need this log here, since the `controller.load_napp` will do the proper logging procedure...
May be just a debug message.

diraol commented on kytos/core/napps/napps_dir_listener.py:
```diff
@@ -0,0 +1,68 @@
+                observer.
+        """
+        napp = self._get_napp(event.src_path)
+        log.info('The NApp "%s/%s" was disabled.', *napp)
```
I don't think we need this log here, since the `controller.load_napp` will do the proper logging procedure...
May be just a debug message.

diraol commented on tests/test_core/test_napp_dir_listener.py:
```diff
@@ -0,0 +1,26 @@
+
+    def test_on_created(self):
+        """Test whether on_created is calling load_napp."""
+        self.napp_dir_listener.on_created(self.event)
```
@cemsbr is this test ok? Shouldn't we test this by creating a directory under the 'napps_dir' (mocked) to see the event being dispatched by the observer?

cemsbr replied:
We can make tests simpler by assuming watchdog works and that it will call the appropriate method.

renanrodrigo committed:
Make the watchdog ignore NApp installing

As NApps are installed in a subdirectory of the napps_dir, we need to tell the watchdog observer to only look for changes when enabling NApps.

diraol commented on kytos/core/napps/napps_dir_listener.py:
```diff
@@ -0,0 +1,72 @@
+        Args:
+            controller(kytos.core.controller): A controller instance.
+        """
+        super(NAppDirListener, self).__init__()
```
Python new syntax....
```python
super().__init__()
```

diraol commented:
And do not forget to make the PR to both branches (``master`` and ``2017.1rc``)

cemsbr reviewed (changes requested):
To be consistent, rename napps_dir_listener.py to napp_dir_listener.py (singular as in the class name).

macartur committed:
Small changes

- Update super method to use super().__init__()
- Changing the file napps_dir_listener.py to napp_dir_listener.py
